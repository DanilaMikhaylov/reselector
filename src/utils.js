'use strict'

const t = require('@babel/types')
const hash = require('string-hash')
const path = require('path')

/**
 * Basic React.Fragment check
 * We can improve it by checking import aliases if needs
 */
const isReactFragment = (node) => {
  if (!node) return false

  if (t.isJSXFragment(node)) {
    return true
  }

  const [element] = node.arguments || []

  if (t.isIdentifier(element)) {
    return element.name === 'Fragment'
  }

  if (t.isMemberExpression(element)) {
    return element.object.name === 'React' && element.property.name === 'Fragment'
  }

  return false
}

const isReactElement = (node) => {
  if (t.isJSXElement(node)) {
    return true
  }

  const { callee } = node

  if (!t.isMemberExpression(callee)) return false

  const { object, property } = callee

  return object.name === 'React' && property.name === 'createElement'
}

const isElement = node => isReactElement(node) && !isReactFragment(node)

const projectPath = process.cwd()

const getNode = (p) => {
  const { parent } = p

  if (isReactFragment(parent)) {
    return getNode(p.parentPath)
  }

  let isComponent = false

  if (![
    t.isConditionalExpression,
    t.isLogicalExpression,
    t.isReturnStatement,
    t.isArrowFunctionExpression,
  ].some(f => f(parent))) {
    return null
  }

  let currentPath = p
  const prevPaths = []

  do {
    isComponent = isComponent || (
      t.isArrowFunctionExpression(currentPath.node)
      || t.isFunctionExpression(currentPath.node)
      || t.isClassExpression(currentPath.node)
    )

    prevPaths.push(currentPath)
    currentPath = currentPath.parentPath

    if (!currentPath) {
      break
    }

    let componentNode = currentPath.node

    switch (currentPath.type) {
      case 'VariableDeclaration': {
        [componentNode] = currentPath.node.declarations

        if (!isComponent) {
          break
        }
      }
      // falls through
      case 'ClassDeclaration':
      case 'ExportDefaultDeclaration':
      case 'ExportNamedDeclaration':
      case 'FunctionDeclaration': {
        if (currentPath.parent.type === 'ExportNamedDeclaration') {
          prevPaths.push(currentPath)
          currentPath = currentPath.parentPath
        }

        return { rootPath: currentPath, prevPaths, componentNode }
      }
    }

    if (isElement(componentNode)) {
      break
    }
  } while (currentPath)

  return null
}

const getName = ({ rootPath, componentNode }) =>
  (rootPath.type === 'ExportDefaultDeclaration' ||
  rootPath.parent.type === 'ExportDefaultDeclaration'
    ? 'default'
    : componentNode.id.name)

const getId = (filename, name) =>
  hash(`${path.relative(projectPath, filename)}:${name}`).toString(16)

module.exports = {
  getNode,
  getName,
  getId,
  isElement,
  isReactFragment,
  isReactElement,
}
