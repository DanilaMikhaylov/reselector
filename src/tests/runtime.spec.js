import { resolveBy, select } from '../'
import { getHashmapFromComment, buildComment } from '../utils'
import * as app from './App'

const TEST_ID = 'data-tid'

const resolve = resolveBy(require.resolve)

describe('runtime', () => {
  it('should resolve components with right test ids', () => {
    const exports = resolve('./App')

    Object.entries(app).forEach(([key, value]) => {
      expect(exports[key][TEST_ID]).toEqual(value[TEST_ID])
    })
  })

  it('should provide an access for unexported components', () => {
    const { UnexportComponent } = resolve('./App')

    expect(select`${UnexportComponent}`).toMatchSnapshot()
  })

  it('should throw an error for undefined component', () => {
    const { UndefinedComponent } = resolve('./App')

    expect(() => select`${UndefinedComponent}`).toThrowErrorMatchingSnapshot()
  })

  it('should get hashmap from comment', () => {
    const hashmap = { Button: { [TEST_ID]: 'button-hash' } }
    expect(getHashmapFromComment(`/*${buildComment(hashmap)}*/`)).toEqual(hashmap)
  })
})
