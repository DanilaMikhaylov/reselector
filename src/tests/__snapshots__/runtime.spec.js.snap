// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime should provide an access for unexported components 1`] = `"[data-tid~=\\"2cf9b58f\\"]"`;

exports[`runtime should throw an error for undefined component 1`] = `"reselector: you can't use undefined value in select"`;
